### PR TITLE
Schedule appointment only when exiting patient summary

### PR DESCRIPTION
**Pull request by pratul**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt [hunk 1]
```diff
@@ -32,11 +32,16 @@ class BloodPressureEntrySheet : BottomSheetActivity() {
...
   companion object {
     private const val KEY_PATIENT_UUID = "patientUuid"
+    private const val EXTRA_WAS_BP_SAVED = "wasBpSaved"
 
-    fun intent(context: Context, patientUuid: UUID): Intent {
+    fun startForResult(activity: TheActivity, context: Context, patientUuid: UUID, requestCode: Int) {
       val intent = Intent(context, BloodPressureEntrySheet::class.java)
       intent.putExtra(KEY_PATIENT_UUID, patientUuid)
-      return intent
+      activity.startActivityForResult(intent, requestCode)
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Calling screen should be responsible for starting intents. The incoming screen shouldn't know about the `requestCode`. Can you move `startActivityForResult()` to summary screen?

File: app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt [hunk 2]
```diff
@@ -93,8 +98,11 @@ class BloodPressureEntrySheet : BottomSheetActivity() {
     diastolicEditText.requestFocus()
   }
 
-  fun finishAndScheduleAppointment(patientUuid: UUID) {
-    startActivity(ScheduleAppointmentSheet.intent(this, patientUuid))
+  fun finishAfterBloodPressureSaved() {
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/bp/entry/BloodPressureEntrySheet.kt]: Hey… this function name is no longer relevant right? It now finishes immediately.

File: app/src/main/java/org/simple/clinic/home/overdue/OverdueListAdapter.kt [hunk 3]
```diff
@@ -78,12 +74,14 @@ class OverdueListViewHolder(
...
 
-  private val patientName by bindView<TextView>(R.id.overdue_patient_name_gender_age)
+  private val patientName by bindView<TextView>(R.id.overdue_patient_name_gender)
   private val patientBP by bindView<TextView>(R.id.overdue_patient_bp)
   private val overdueDays by bindView<TextView>(R.id.overdue_days)
   private val callButton by bindView<ImageButton>(R.id.overdue_patient_call)
   private val separator by bindView<View>(R.id.overdue_separator)
   private val actionsContainer by bindView<LinearLayout>(R.id.overdue_actions_container)
+  private val age by bindView<TextView>(R.id.overdue_patient_age)
+  private val phoneNumber by bindView<TextView>(R.id.overdue_patient_phoneNo)
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/home/overdue/OverdueListAdapter.kt]: Whoops, I just realized that you're not following the naming convention for View fields here. Please fix them. :)

File: app/src/main/java/org/simple/clinic/home/overdue/OverdueListAdapter.kt [hunk 4]
```diff
@@ -100,22 +98,15 @@ class OverdueListViewHolder(
...
-    val spanBuilder = Truss()
-        .append(appointment.name)
-        .append(" ")
-        .pushSpan(ForegroundColorSpan(ContextCompat.getColor(context, R.color.black_opacity_50)))
-        .pushSpan(AbsoluteSizeSpan(14, true))
-        .append(context.getString(R.string.overdue_list_item_patient_gender_age, context.getString(appointment.gender.displayTextRes), appointment.age))
-        .popSpan()
 
-    patientName.text = spanBuilder.build()
+    patientName.text = """${appointment.name}, ${context.getString(appointment.gender.displayLetterRes)}"""
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/home/overdue/OverdueListAdapter.kt]: `"""` is useful for mult-line strings. Did you intentionally add this?

File: app/src/main/java/org/simple/clinic/home/overdue/OverdueListAdapter.kt [hunk 5]
```diff
@@ -131,6 +122,10 @@ class OverdueListViewHolder(
       callButton.visibility = View.VISIBLE
       separator.visibility = View.VISIBLE
     }
+
+    age.text = """${appointment.age}"""
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/home/overdue/OverdueListAdapter.kt]: Same question about `"""` here

File: app/src/main/java/org/simple/clinic/patient/Gender.kt [hunk 6]
```diff
@@ -5,16 +5,16 @@ import com.squareup.moshi.Json
 import org.simple.clinic.R
 import org.simple.clinic.util.RoomEnumTypeConverter
 
-enum class Gender(@StringRes val displayTextRes: Int) {
+enum class Gender(@StringRes val displayTextRes: Int, val displayLetterRes: Int) {
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/patient/Gender.kt]: Can you also annotate this new param with `@StringRes`?

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt [hunk 7]
```diff
@@ -125,15 +125,61 @@ class PatientSummaryScreenController @Inject constructor(
...
-  private fun handleBackClicks(events: Observable<UiEvent>): Observable<UiChange> {
-    return events
+  private fun backAndDoneClicks(events: Observable<UiEvent>): Observable<UiChange> {
+    val patientUuids = events
+        .ofType<PatientSummaryScreenCreated>()
+        .map { it.patientUuid }
+
+    val bloodPressureSaves = events
+        .ofType<PatientSummaryBloodPressureSaved>()
+        .startWith(PatientSummaryBloodPressureSaved(false))
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt]: Hrm, `PatientSummaryBloodPressureSaved` sounds like an event that is emitted when a BP is saved. Can we rename this to `PatientSummaryBloodPressureClosed` or something?

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt [hunk 8]
```diff
@@ -125,15 +125,61 @@ class PatientSummaryScreenController @Inject constructor(
         .map { (_, patientUuid) -> { ui: Ui -> ui.showUpdatePrescribedDrugsScreen(patientUuid) } }
   }
 
-  private fun handleBackClicks(events: Observable<UiEvent>): Observable<UiChange> {
-    return events
+  private fun backAndDoneClicks(events: Observable<UiEvent>): Observable<UiChange> {
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt]: This is another thing that would have been covered already if we had a guidelines wiki in page. We've so far named ui change functions according to the action they're performing and not the events they're consuming. Can you rename both `backAndDoneClicks` and `scheduleAppointmentSheetCloses`?

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt (hunk 7 above)
Comment (sanchita) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreenController.kt]: ```PatientSummaryBloodPressureSaved``` is an event that is emitted only when Blood pressure is saved.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt [hunk 9]
```diff
@@ -95,7 +101,13 @@ class PatientSummaryScreen(context: Context, attrs: AttributeSet) : RelativeLayo
...
+    Observable.mergeArray(
+        screenCreates(),
+        backClicks(),
+        doneClicks(),
+        adapterUiEvents,
+        bloodPressureSaves(),
+        appointmentScheduleSheetClosed())
         .observeOn(io())
         .compose(controller)
         .observeOn(mainThread())
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt]: Formatting: the `mergeArray()` operators have the same indentation as `observeOn()` and other operators, making it look like they're at the same nesting level. 

One trick I use is to move the static function in a new line too, making the nesting clear:

```kotlin
Observable
    .mergeArray(
        screenCreates(),
        backClicks(),
        doneClicks(),
        adapterUiEvents,
        bloodPressureSaves(),
        appointmentScheduleSheetClosed())
    .observeOn(io())
    .compose(controller)
    .observeOn(mainThread())
    .takeUntil(RxView.detaches(this))
    .subscribe { uiChange -> uiChange(this) }
```

It's okay if you want to change this in your next PR.

File: app/src/main/java/org/simple/clinic/scheduleappointment/ScheduleAppointmentSheet.kt [hunk 10]
```diff
@@ -24,10 +25,10 @@ class ScheduleAppointmentSheet : BottomSheetActivity() {
   companion object {
     private const val KEY_PATIENT_UUID = "patientUuid"
 
-    fun intent(context: Context, patientUuid: UUID): Intent {
+    fun startForResult(activity: TheActivity, context: Context, patientUuid: UUID, requestCode: Int) {
       val intent = Intent(context, ScheduleAppointmentSheet::class.java)
       intent.putExtra(KEY_PATIENT_UUID, patientUuid)
-      return intent
+      activity.startActivityForResult(intent, requestCode)
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/scheduleappointment/ScheduleAppointmentSheet.kt]: Looks like you forgot to revert this :)

**Pull request merged**